### PR TITLE
Ui components

### PR DESCRIPTION
**Pull request by JulienPradet**
That's a bit weird because of the coupling of this with Storybook. But it makes sense in term of development workflow. Don't know though how it will go with the actual storybook/styleguide usage documentation.

### CODE REVIEW

File: source/docs/essentials/create-a-ui-component.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+folder to find examples into our source code.
+
+But first, let's define what is an ideal UI Component.
+
+## The ideal UI Component
+
+In Front-Commerce we call UI component any component that is:
+
+- **Reusable in many contexts** If a component is used only once in the whole
+  application, it might be the sign that it does not exist purely for UI
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
  application, it might be a smell that it does not exist purely for UI
```

I am not sure about this one, but as a developer « smell » gives me a direct context and more information that this is technical debt or incorrect abstraction.

File: source/docs/essentials/create-a-ui-component.md [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+## How to build a UI Component?
+
+Alright, that's nice in theory, but how does it translate in practice? We'll try
+to get a bit more tangible by creating a UI component needed for adding a
+Reinsurance Banner in a page.
+
+![The reinsurance banner that we will implement](assets/reinsurance.jpg)
+
+### Step 1: Defining the components
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]: Other pages do not have the "Step x" prefix. I’d suggest removing them here for consistency.

```suggestion
### Defining the components
```

File: source/docs/essentials/create-a-ui-component.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+will depend on your project and your team. It may be a better idea to keep
+things simple. It's often easier to wait and see how it can be reused later.
+</blockquote>
+
+We won't be able to detail each component here. We will focus on
+`theme/components/molecules/IllustratedContent` instead. But keep in mind that
+any UI component in Front-Commerce will look similar to what we are going to
+build here.
+
+### Step 2: Setup your dev environment
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
### Setup your dev environment
```

File: source/docs/essentials/create-a-ui-component.md [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+<figcaption style="margin: 0.5em; text-align: center">Take a look at this magical development playground!</figcaption>
+</figure>
+
+> Learn more:
+>
+> - about Storybook itself by reading the
+>   [official Storybook documentation](https://storybook.js.org/basics/introduction/)
+> - about [our Storybook usage](#TODO) by reading our documentation
+
+### Step 3: Implement your component
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
### Implement your component
```

File: source/docs/essentials/create-a-ui-component.md [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+- document edge cases
+- provide a test suite thanks to snapshot testing
+  ([Storyshots](https://github.com/storybooks/storybook/tree/master/addons/storyshots))
+- create a common discussion base for designers, product managers, marketers,
+  etc.
+
+[Learn more about Storybook.](#TODO)
+
+### Step 4: Use the component
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
### Use the component
```

File: source/docs/essentials/create-a-ui-component.md [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    // here since this component will be heavily
+    // used through your application
+  };
+
+  export default IllustratedContent;
+  ```
+
+- `my-module/web/theme/components/molecules/IllustratedContent/index.js`: will proxy the
+  IllustratedContent.js file in order to be able to do imports on the folder
+  directly.
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
  directly. See [this blog post](http://bradfrost.com/blog/post/this-or-that-component-names-index-js-or-component-js/) for more context about this pattern.
```

It could be a blockquote too if it seems better.

File: source/docs/essentials/create-a-ui-component.md [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+  width: 30%;
+  max-width: 10em;
+}
+```
+
+If it didn't reload properly, it is because you need to restart the application
+every time you override a component in order to let the application know that the
+file location it should load has changed. But note that there is an upcoming
+improvement [#63](https://gitlab.com/front-commerce/front-commerce/issues/63) that
+should make things for you in the future.
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
should remove the need for a restart in the future.
```

File: source/docs/essentials/create-a-ui-component.md [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+
+As a side note, we also use [BEM convention](http://getbem.com/naming/) for our
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]: We should not forget to add a link to our future « Advanced > Theme > BEM » page here when it will be live.

How to keep track of it?

File: source/docs/essentials/create-a-ui-component.md [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+If you feel the need to understand why we went for this organization, feel
+free to refer to [React components structure](#TODO)
+first.
+</blockquote>
+
+In this documentation, you will learn how to build a UI Component. We will do so by
+creating our own. Front-Commerce’s core UI components follow the same principles
+and you could dive into the `node_modules/front-commerce/src/web/theme/components`
+folder to find examples into our source code.
+
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]: We could add a sentence here explaining that we chose to also keep Storybook mentioned in the process because it is for us part of the same workflow and link to the other page when it is done so people could have some background.

File: source/docs/essentials/create-a-ui-component.md (hunk 8 above)
Comment (JulienPradet) [reply source/docs/essentials/create-a-ui-component.md]: I'll add a TODO comment.

File: source/docs/essentials/create-a-ui-component.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+But first, let's define what is an ideal UI Component.
+
+## The ideal UI Component
+
+In Front-Commerce we call UI component any component that is:
+
+- **Reusable in many contexts** If a component is used only once in the whole
+  application, it might be a smell that it does not exist purely for UI
+  purpose. The component most likely needs to be moved to the `theme/modules` folder.
```
Comment (JulienPradet) [inline source/docs/essentials/create-a-ui-component.md]: Add link to Business components doc

File: source/docs/essentials/create-a-ui-component.md [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+- create a common discussion base for designers, product managers, marketers,
+  etc.
+
+[Learn more about Storybook.](#TODO)
+
+### Use the component
+
+Once you are satisfied with your component, you can use it anywhere. In this case,
+the `IllustratedContent` was to be used in the Reinsurance Banner. Thus, this
+module's component would look like this:
```
Comment (JulienPradet) [inline source/docs/essentials/create-a-ui-component.md]: Add a mention about why you can load it directly from `theme`.

File: source/docs/essentials/create-a-ui-component.md [hunk 12]
```diff
@@ -0,0 +1,327 @@
+---
+id: create-a-ui-component
+title: Create a UI Component
+---
+
+In Front Commerce we split our components in two categories: the **UI**
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
In Front-Commerce components are classified under two categories: the **UI**
```

(not sure about this one)

File: source/docs/essentials/create-a-ui-component.md [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+first.
+</blockquote>
+
+In this documentation, you will learn how to build a UI Component. We will do so by
+creating our own. We will use [Storybook](#TODO) in the process because it is our
+usual workflow when creating a UI Component. But if you don't need it or prefer to
+add your stories later, feel free to leave the parts mentioning Storybook for later.
+
+Front-Commerce’s core UI components follow the same principles
+and you could dive into the `node_modules/front-commerce/src/web/theme/components`
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
and you could dive into the [`node_modules/front-commerce/src/web/theme/components`](https://gitlab.com/front-commerce/front-commerce/tree/develop/src/web/theme/components)
```

File: source/docs/essentials/create-a-ui-component.md [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+folder to find examples into our source code.
+
+But first, let's define what is an ideal UI Component.
+
+## The ideal UI Component
+
+In Front-Commerce we call UI component any component that is:
+
+- **Reusable in many contexts** If a component is used only once in the whole
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
- **Reusable in many contexts**: if a component is used only once in the whole
```

File: source/docs/essentials/create-a-ui-component.md [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+  application, it might be a smell that it does not exist purely for UI
+  purpose. The component most likely needs to be moved to the `web/theme/modules` folder. <!-- TODO link to the business component page -->
+  That's also the reason why we avoid to give names too close to its business
+  use. For instance, we don't want to have a UI component that would be called
+  `ProductDescription`. It would be better to go for a `Description` that would
+  thus be reusable by a Category.
+- **Focused on abstracting away UI concerns** The goal of UI components is to
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
- **Focused on abstracting away UI concerns**: the goal of UI components is to
```

File: source/docs/essentials/create-a-ui-component.md [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+  hide styles or DOM concerns from their parents. It may be hard to achieve
+  sometimes, but it will greatly improve the parent component's readability. For
+  instance, a UI component should not give the opportunity to pass a `className`
+  in its props as it may lead to many styles inconsistencies across the theme.
+
+## How to build a UI Component?
+
+Alright, that's nice in theory, but how does it translate in practice? We'll try
+to get a bit more tangible by creating a UI component needed for adding a
+Reinsurance Banner in a page.
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
Reinsurance Banner in a page similar to the following mockup.
```

File: source/docs/essentials/create-a-ui-component.md [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+
+![The reinsurance banner that we will implement](assets/reinsurance.jpg)
+
+### Defining the components
+
+First, let's split the mockup in several UI components following the [Atomic Design Principles](http://atomicdesign.bradfrost.com/).
+
+![The various components that will make up our banner](assets/reinsurance-with-areas.jpg)
+
+- **`web/theme/components/atoms/typography/Heading`:** enforces consistent font sizes in our theme
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
- **`web/theme/components/atoms/typography/Heading` (green):** enforces consistent font sizes in our theme
```

File: source/docs/essentials/create-a-ui-component.md [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+  for any title
+- **`web/theme/components/atoms/Icon`:** enforces icon sizes and accessibility guidelines
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
- **`web/theme/components/atoms/Icon` (purple):** enforces icon sizes and accessibility guidelines
```

File: source/docs/essentials/create-a-ui-component.md [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+- **`web/theme/components/molecules/IllustratedContent`:** displays some content illustrated by an
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
- **`web/theme/components/molecules/IllustratedContent` (red):** displays some content illustrated by an
```

File: source/docs/essentials/create-a-ui-component.md [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+  image and aligns images consistently across the whole theme
+- **`web/theme/components/organisms/InlineCards`:** manages a list of cards and inlines them,
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
- **`web/theme/components/organisms/InlineCards` (orange):** manages a list of cards and inlines them,
```

File: source/docs/essentials/create-a-ui-component.md [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+```
+
+If it didn't reload properly, it is because you need to restart the application
+every time you override a component in order to let the application know that the
+file location it should load has changed. But note that there is an upcoming
+improvement [#63](https://gitlab.com/front-commerce/front-commerce/issues/63) that
+should remove the need for a restart in the future.
+
+<!-- TODO: add a link to our Advanced/Theme/BEM article when it's done. -->
+As a side note, we also use [BEM convention](http://getbem.com/naming/) for our
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]: It would be good to wrap this paragraph in a `<blockquote class="note">` IMO

File: source/docs/essentials/create-a-ui-component.md [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+As you can see, we did not use a relative import. This is because in Front-Commerce
+we have a few aliases that will let you import files without worrying about your
+current position in the folder structure.
+
+In our case, if the `ReinsuranceBanner` was in
+`my-module/web/theme/modules/ReinsuranceBanner`, we don't have to import the
+`IllustratedContent` by using relative paths
+`../../components/molecules/IllustratedContent` but we can remain with
+`theme/components/molecules/IllustratedContent` which is more explicit. This is
+possible for any file located in the folder `web/theme` of a module.
```
Comment (real34) [inline source/docs/essentials/create-a-ui-component.md]:
```suggestion
possible for any file located in the folder `web/theme` of a module. <-- TODO Add a link to the aliases documentation page -->
```

I guess we will explain it more in depth at some point in the advanced section of the documentation.
